peer/peer.go

```diff
@@ -0,0 +1,2114 @@
+	// and is used to assign an id to a peer.
+	nodeCount int32
+
+	// zeroHash is the zero value hash (all zeros).  It is defined as a
+	// convenience.
+	zeroHash wire.ShaHash
+)
+
+// Config is the struct to hold configuration options useful to Peer
+type Config struct {
+	// SOCKS5 proxy (eg. 127.0.0.1:9050) to use for connections
+	Proxy string
+
+	// Whether to use the imulation test network
+	SimNet bool
+
+	// Whether to use the regression test network
+	RegressionTest bool
+
+	// Disable listening for incoming connections
+	// Listening is automatically disabled if the --connect or --proxy options
+	// are used without also specifying listen interfaces via --listen
+	DisableListen bool
+
+	// If non-nil, the callback to be invoked when reading a peer message.
+	OnRead func(int, *wire.Message, error)
+
+	// If non-nil, the callback to be invoked when writing a peer message.
+	OnWrite func(int, *wire.Message, error)
+
+	// User agent string to be used in peer messages.
+	UserAgentName string
+
+	// User agent version to be used in peer messages.
+	UserAgentVersion string
+}
+
+// minUint32 is a helper function to return the minimum of two uint32s.
+// This avoids a math import and the need to cast to floats.
+func minUint32(a, b uint32) uint32 {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+// newNetAddress attempts to extract the IP address and port from the passed
+// net.Addr interface and create a bitcoin NetAddress structure using that
+// information.
+func newNetAddress(addr net.Addr, services wire.ServiceFlag) (*wire.NetAddress, error) {
+	// addr will be a net.TCPAddr when not using a proxy.
+	if tcpAddr, ok := addr.(*net.TCPAddr); ok {
+		ip := tcpAddr.IP
+		port := uint16(tcpAddr.Port)
+		na := wire.NewNetAddressIPPort(ip, port, services)
+		return na, nil
+	}
+
+	// addr will be a socks.ProxiedAddr when using a proxy.
+	if proxiedAddr, ok := addr.(*socks.ProxiedAddr); ok {
+		ip := net.ParseIP(proxiedAddr.Host)
+		if ip == nil {
+			ip = net.ParseIP("0.0.0.0")
+		}
+		port := uint16(proxiedAddr.Port)
+		na := wire.NewNetAddressIPPort(ip, port, services)
+		return na, nil
+	}
+
+	// For the most part, addr should be one of the two above cases, but
+	// to be safe, fall back to trying to parse the information from the
+	// address string as a last resort.
+	host, portStr, err := net.SplitHostPort(addr.String())
+	if err != nil {
+		return nil, err
+	}
+	ip := net.ParseIP(host)
+	port, err := strconv.ParseUint(portStr, 10, 16)
+	if err != nil {
+		return nil, err
+	}
+	na := wire.NewNetAddressIPPort(ip, uint16(port), services)
+	return na, nil
+}
+
+// outMsg is used to house a message to be sent along with a channel to signal
+// when the message has been sent (or won't be sent due to things such as
+// shutdown)
+type outMsg struct {
+	msg      wire.Message
+	doneChan chan struct{}
+}
+
+// Stats is the collection of stats related to a peer
+type Stats struct {
```

Comment:
This structure is not used outside of the peer package and therefore should not be exported.